design/one-pager-moving-packages.md

```diff
@@ -0,0 +1,462 @@
+
+```yaml
+apiVersion: meta.pkg.crossplane.io/v1
+kind: Provider
+metadata:
+  name: provider-family-aws
+spec:
+  replaces:
+  - xpkg.upbound.io/upbound/provider-family-aws
+```
+
+This tells Crossplane that the `crossplane-contrib/provider-family-aws` package
+replaces the `upbound/provider-family-aws` package.
+
+When the revision controller encounters a package with a replaces stanza, it'll
+deactivate the replaced package(s) before installing the new package.
+
+When the replaced package is deactivated, the new package can be installed
+without a conflict. A deactivated package [owns][owner-ref] but doesn't
+[control][controller-ref] its payload (e.g. a Provider's MR CRDs). This
+eliminates the controller reference conflict.
+
+A deactivated package exists, but in an inactive state. Specifically:
+
+* It appears under (e.g.) `kubectl get providers`
+* At least one revision appears under (e.g.) `kubectl get providerrevisions`
+* The package's `spec.revisionActivationPolicy` is set to `Manual`
+* All revisions have `spec.desiredState` set to `Inactive`
+* It has no runtime - e.g. no Deployment of a provider's runtime binary
+* It owns but doesn't control its package payload - e.g. a provider's MR CRDs
+
+Crucially, deactivating a package doesn't remove its payload. For example
+deactivating a provider doesn't remove its MR CRDs like uninstalling it would.
+The new package shares ownership (in the [owner reference][owner-ref] sense)
+with the replaced package, until the replaced package is manually uninstalled.
+This allows the package manager to seamlessly handover ownership and control of
+a package's payload.
+
+To deactivate a replaced packaged, the revision controller will:
+
+1. Map the replaced source to its active package revision using the Lock
+2. Map the package revision to its package using its owner reference
+3. Set the package's `spec.revisionActivationPolicy` to `Manual`
+4. Set the package revision's `spec.desiredState` to `Inactive`
+
+One the replace source no longer appears in the Lock, the revision controller
+will proceed to install the new package.
+
+Dependency installation is ultimately handled by the revision controller - see
+the sequence diagram in [the background section](#background). Handling the
+`replaces` stanza in the revision controller means the package manager will
+transparently handle `replaces` for dependencies, not only top level packages.
+
+A package can declare that it replaces another package without the replaced
+package's consent. This means a malicious package could replace another package
+that wasn't actually moved. Making the `replaces` stanza part of the package
+metadata reduces this risk. Crossplane-aware package registries (e.g.
+https://marketplace.upbound.io) can use the package metadata to prominently
+display the fact that a package will replace another package. This will help
+avoid unintentionally installing a package that replaces another.
```

Comment:
Showing a warning in the registry sounds good, but this may not be possible/feasible to warn if one of the dependencies of an installed package has a replacement, e.g. registry parses all dependency tree and "one of the dependencies contains a replacement". 

At least some sort of admission/approval in the Crossplane-aware package registries, before accepting a package with a replace stanza could be helpful.

Reply:
Agreed, but I don't think I'll go into more detail on it in this document given that it's more of a registry concern.